# Tasks 8–19: Missing Values, Duplicates, and Data Cleaning

In [1]:
import numpy as np
import pandas as pd

df = pd.read_csv("../Data/Titanic-Dataset.csv")

In [2]:
# Task 8: Count non-missing values
df.notnull().sum()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

In [3]:
# Task 9: Check duplicate rows
print("Duplicate Rows:", df.duplicated().sum())

Duplicate Rows: 0


**Student Question (Task 9):**
Duplicate rows = 0

In [4]:
# Task 10: Check target variable values
df["Survived"].unique()

array([0, 1])

In [5]:
# Task 11: Count unique values in every column
df.nunique()

PassengerId    891
Survived         2
Pclass           3
Name           891
Sex              2
Age             88
SibSp            7
Parch            7
Ticket         681
Fare           248
Cabin          147
Embarked         3
dtype: int64

In [6]:
# Task 12: Descriptive statistics
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


**Student Questions (Task 12):**
1. Average passenger age ≈ 29.70 years
2. Minimum passenger age = 0.42 years
3. Maximum passenger age = 80 years
4. Average fare ≈ 32.20
5. Maximum fare ≈ 512.33
6. Median age = 28 years

**Observation:**
The average age of passengers is approximately 29.7 years, with most falling between 20 and 38 (25th–75th percentile). Passenger fares range from 0 to 512.33 — a huge spread. The large gap between the median fare (14.45) and the maximum (512.33) suggests a small number of passengers paid dramatically more than most, likely first-class or luxury cabins.

In [7]:
# Task 13: Count missing values
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [8]:
# Task 14: Show only columns with missing values, sorted
missing_values = df.isnull().sum()
missing_values[missing_values > 0].sort_values(ascending=False)

Cabin       687
Age         177
Embarked      2
dtype: int64

In [9]:
# Task 15: Missing value percentage
missing_percentage = (df.isnull().sum() / len(df) * 100).round(2)
pd.DataFrame({
    "Missing Count": df.isnull().sum(),
    "Missing Percentage": missing_percentage
}) 

,Missing Count,Missing Percentage
PassengerId,0,0.00
Survived,0,0.00
Pclass,0,0.00
Name,0,0.00
Sex,0,0.00
Age,177,19.87
SibSp,0,0.00
Parch,0,0.00
Ticket,0,0.00
Fare,0,0.00


**Student Questions (Tasks 14–15):**
1. Column with largest number of missing values: Cabin (687 missing)
2. Highest missing percentage: Cabin ≈ 77.10%
3. Age missing percentage ≈ 19.87% — significant, not small
4. Embarked missing percentage ≈ 0.22% — very minor
5. Filling Cabin would not be reliable — with over three-quarters of values missing, any imputation (mean/mode/constant) would be mostly guessed data and could seriously bias analysis. Best to leave it unchanged or drop it if unusable later.


In [10]:
# Task 16: Remove duplicate rows
df = df.drop_duplicates().copy()

In [11]:
# Task 17: Handle missing Age values with median
print("Median Age:", df["Age"].median())
df["Age"] = df["Age"].fillna(df["Age"].median())

Median Age: 28.0


In [12]:
# Task 18: Handle missing Embarked values with mode
print("Mode of Embarked:", df["Embarked"].mode()[0])
df["Embarked"] = df["Embarked"].fillna(df["Embarked"].mode()[0])

Mode of Embarked: S


In [13]:
# Task 19: Verify missing values after cleaning
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         0
dtype: int64

**Student Questions (Task 19):**
1. Does Age still contain missing values? No — 0 missing after median fill.
2. Does Embarked still contain missing values? No — 0 missing after mode fill.
3. Which column still contains a large number of missing values? Cabin (687 missing, left unchanged as instructed).